src/ciphers/caesar.rs

```diff
@@ -1,43 +1,115 @@
+                    assert_eq!(caesar(&expected, backward_rotation).unwrap(), text);
+                }
+            )*
+        };
+    }
+
+    macro_rules! test_caesar_error_cases {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (text, rotation) = $test_case;
+                    assert_eq!(caesar(&text, rotation), Err(ERROR_MESSAGE));
+                }
+            )*
+        };
     }
 
-    #[test]
-    fn caesar_rot_13() {
-        assert_eq!(caesar("rust", 13), "ehfg");
+    test_caesar_happy_path! {
+        empty_text: ("", 13, ""),
+        rot_13: ("rust", 13, "ehfg"),
+        unicode: ("attack at dawn 攻", 5, "fyyfhp fy ifbs 攻"),
+        rotation_within_alphabet_range: ("Hello, World!", 3, "Khoor, Zruog!"),
+        no_rotation: ("Hello, World!", 0, "Hello, World!"),
+        rotation_at_alphabet_end: ("Hello, World!", 25, "Gdkkn, Vnqkc!"),
+        longer: ("The quick brown fox jumps over the lazy dog.", 5, "Ymj vznhp gwtbs ktc ozrux tajw ymj qfed itl."),
+        non_alphabetic_characters: ("12345!@#$%", 3, "12345!@#$%"),
+        uppercase_letters: ("ABCDEFGHIJKLMNOPQRSTUVWXYZ", 1, "BCDEFGHIJKLMNOPQRSTUVWXYZA"),
+        mixed_case: ("HeLlO WoRlD", 7, "OlSsV DvYsK"),
+        with_whitespace: ("Hello, World!", 13, "Uryyb, Jbeyq!"),
+        with_special_characters: ("Hello!@#$%^&*()_+World", 4, "Lipps!@#$%^&*()_+Asvph"),
+        with_numbers: ("Abcd1234XYZ", 10, "Klmn1234HIJ"),
     }
 
-    #[test]
-    fn caesar_unicode() {
-        assert_eq!(caesar("attack at dawn 攻", 5), "fyyfhp fy ifbs 攻");
+    test_caesar_error_cases! {
+        negative_rotation: ("Hello, World!", -5),
+        empty_input_negative_rotation: ("", -1),
+        empty_input_large_rotation: ("", 27),
+        large_rotation: ("Large rotation", 139),
+        large_rotation_with_big_input: ("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ac ultrices ante, at gravida ante. Quisque luctus, ligula nec dictum facilisis, elit leo luctus arcu, ut auctor sapien turpis ut mauris. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nulla vel orci sit amet sem efficitur sagittis a quis augue. Donec semper quam tincidunt hendrerit cursus. Duis placerat gravida diam, in interdum purus dapibus in.", 139),
+        very_large_rotation_with_big_input: ("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ac ultrices ante, at gravida ante. Quisque luctus, ligula nec dictum facilisis, elit leo luctus arcu, ut auctor sapien turpis ut mauris. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nulla vel orci sit amet sem efficitur sagittis a quis augue. Donec semper quam tincidunt hendrerit cursus. Duis placerat gravida diam, in interdum purus dapibus in.", 345876),
```

Comment:
Why do you insist on these two?

Reply:
Because the previous implementation failed to pass these two tests

Reply:
Was is failing due to the length of the input or size of the rotation?

Reply:
Rotation size

Reply:
Ok, removed